### PR TITLE
Verify heap integrity after sweep.

### PR DESCRIPTION
**Pull request by erikcorry**
Also removes a free area zapping function that did
not yield consistent heaps.

### CODE REVIEW

**Comment by erikcorry** [comment]
I changed some indentation, you might want to use "hide whitespace" view.

File: src/third_party/dartino/gc_metadata.cc [hunk 1]
```diff
@@ -208,7 +208,7 @@ GcMetadata::Destination GcMetadata::calculate_object_destinations(
       uint32* overhang_bits =
           mark_bits_for(end_of_last_source_object_moved - WORD_SIZE);
       ASSERT((*overhang_bits & 1) != 0);
-      *overhang_bits &= ~((1 << overhang) - 1);
+      *overhang_bits &= ~((1u << overhang) - 1);
```
Comment (kasperl) [inline src/third_party/dartino/gc_metadata.cc:211]: Nit: I'd use 1U to make it consistent with 1L where a lower case l would suck.

File: src/third_party/dartino/gc_metadata.cc [hunk 2]
```diff
@@ -274,13 +274,13 @@ void GcMetadata::slow_mark(HeapObject* object, uword size) {
   ASSERT(words + mask_shift > 32);
   for (words -= 32 - mask_shift; words >= 32; words -= 32)
     *bits++ = 0xffffffffu;
-  *bits |= (1 << words) - 1;
+  *bits |= (1u << words) - 1;
```
Comment (kasperl) [inline src/third_party/dartino/gc_metadata.cc:277]: Nit: I'd use 1U to make it consistent with 1L where a lower case l would suck.

File: src/third_party/dartino/gc_metadata.cc [hunk 3]
```diff
@@ -274,13 +274,13 @@ void GcMetadata::slow_mark(HeapObject* object, uword size) {
...
 }
 
 void GcMetadata::mark_stack_overflow(HeapObject* object) {
   uword address = object->_raw();
   uint8* overflow_bits = overflow_bits_for(address);
-  *overflow_bits |= 1 << ((address >> CARD_SIZE_LOG_2) & 7);
+  *overflow_bits |= 1u << ((address >> CARD_SIZE_LOG_2) & 7);
```
Comment (kasperl) [inline src/third_party/dartino/gc_metadata.cc:283]: Nit: I'd use 1U to make it consistent with 1L where a lower case l would suck.

File: src/third_party/dartino/object_memory_mark_sweep.cc [hunk 4]
```diff
@@ -574,87 +533,135 @@ uword OldSpace::sweep() {
...
+      // We set the object starts for this card to NO_OBJECT_START, but
+      // that's not very helpful.  Repair it to point to the end of the
+      // free area, which is a valid place to iterate from.
+      uint8* end_starts_location = GcMetadata::starts_for(end_of_free);
+      ASSERT(end_of_free < end);
+      *end_starts_location = end_of_free;
+      line += GcMetadata::CARD_SIZE;
+      mark_bits++;
     }
+    end_of_chunk:
```
Comment (kasperl) [inline src/third_party/dartino/object_memory_mark_sweep.cc:608]: Maybe indent one less?

File: src/third_party/dartino/object_memory_mark_sweep.cc [hunk 5]
```diff
@@ -574,87 +533,135 @@ uword OldSpace::sweep() {
...
+void OldSpace::validate_sweep(Chunk* chunk) {
+  uword line = chunk->start();
+  uword object_iterator = line;
+  uword end = line + chunk->size();
+  while (line != end && object_iterator != end - WORD_SIZE) {
+    uint32* mark_bits = GcMetadata::mark_bits_for(object_iterator);
+    uint8* starts = GcMetadata::starts_for(object_iterator);
+    HeapObject* object = HeapObject::from_address(object_iterator);
+    uword size = object->size(program_);
+    bool alive = (*mark_bits & (1 << ((object_iterator - line) / WORD_SIZE))) != 0;
```
Comment (kasperl) [inline src/third_party/dartino/object_memory_mark_sweep.cc:631]: 1U?

File: src/third_party/dartino/object_memory_mark_sweep.cc [hunk 6]
```diff
@@ -574,87 +533,135 @@ uword OldSpace::sweep() {
...
+    ASSERT(GcMetadata::all_mark_bits_are(object, size, alive ? 1 : 0));
+    ASSERT(object->is_a_free_object() == !alive);
+    if (*starts != GcMetadata::NO_OBJECT_START) {
+      ASSERT(*starts < GcMetadata::CARD_SIZE);
+      uword location = *starts | (object_iterator & ~0xffll);
```
Comment (kasperl) [inline src/third_party/dartino/object_memory_mark_sweep.cc:636]: LL

**Review by kasperl — APPROVED** [review]
Looks good to me.

**Pull request merged**